pkg/client/s3/client.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright (c) 2021 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package s3
+
+import (
+	"fmt"
+
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	awscore "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/aws/aws-sdk-go/service/sqs/sqsiface"
+
+	"github.com/triggermesh/aws-event-sources/pkg/apis/sources/v1alpha1"
+	"github.com/triggermesh/aws-event-sources/pkg/aws"
+)
+
+// Client is an alias for the S3API interface.
+type Client = s3iface.S3API
+
+// SQSClient is an alias for the S3API interface.
+type SQSClient = sqsiface.SQSAPI
+
+// ClientGetter can obtain S3 and SQS clients.
+type ClientGetter interface {
+	Get(*v1alpha1.AWSS3Source) (Client, SQSClient, error)
+}
+
+// NewClientGetter returns a ClientGetter for the given secrets getter.
+func NewClientGetter(sg NamespacedSecretsGetter) *ClientGetterWithSecretGetter {
+	return &ClientGetterWithSecretGetter{
+		sg: sg,
+	}
+}
+
+type NamespacedSecretsGetter func(namespace string) coreclientv1.SecretInterface
+
+// ClientGetterWithSecretGetter gets S3 clients using static credentials
+// retrieved using a Secret getter.
+type ClientGetterWithSecretGetter struct {
+	sg NamespacedSecretsGetter
+}
+
+// ClientGetterWithSecretGetter implements ClientGetter.
+var _ ClientGetter = (*ClientGetterWithSecretGetter)(nil)
+
+// Get implements ClientGetter.
+func (g *ClientGetterWithSecretGetter) Get(src *v1alpha1.AWSS3Source) (Client, SQSClient, error) {
+	creds, err := aws.Credentials(g.sg(src.Namespace), &src.Spec.Credentials)
+	if err != nil {
+		return nil, nil, fmt.Errorf("retrieving AWS security credentials: %w", err)
+	}
+
+	sess := session.Must(session.NewSession(awscore.NewConfig().
+		WithRegion(src.Spec.ARN.Region).
+		WithCredentials(credentials.NewStaticCredentialsFromCreds(*creds)),
+	))
```

Comment:
nit - personal preference:

session.Must seems to be doing
```go
func Must(object)
	if err != nil {
		panic(err)
	}
        return object
```

Since an error is already returned from this func, I would use it and let caller decide if to panic.

Reply:
I don't mind changing that, but then I'd do it in all sources in one go.

Reply:
makes sense.